File: orro/cloud.py
"""Tuya Cloud API client."""

from __future__ import annotations

import hashlib
import hmac
import json
import time
from typing import Any
from urllib.parse import urlencode

import requests

from orro.config import Config


def canonical_query(params: dict[str, Any] | None) -> str:
    """Sort and encode query parameters for Tuya API signing."""
    if not params:
        return ""
    pairs: list[tuple[str, str]] = []
    for key in sorted(params):
        value = params[key]
        if value is None:
            continue
        if isinstance(value, (list, tuple)):
            for item in value:
                pairs.append((key, str(item)))
        else:
            pairs.append((key, str(value)))
    return urlencode(pairs)


class TuyaCloud:
    """Tuya Cloud HTTP API client with HMAC-SHA256 request signing."""

    def __init__(self, config: Config) -> None:
        self.endpoint = config.endpoint.rstrip("/")
        self.access_id = config.access_id
        self.access_secret = config.access_secret.encode()
        self.device_id = config.device_id
        self._config = config
        self.access_token: str | None = None

    @staticmethod
    def _body_text(body: Any | None) -> str:
        return "" if body is None else json.dumps(body, separators=(",", ":"))

    def _sign(
        self,
        method: str,
        path: str,
        params: dict[str, Any] | None = None,
        body: Any | None = None,
    ) -> dict[str, str]:
        query = canonical_query(params)
        url_path = path + (f"?{query}" if query else "")
        body_text = self._body_text(body)
        body_hash = hashlib.sha256(body_text.encode()).hexdigest()
        string_to_sign = "\n".join([method.upper(), body_hash, "", url_path])
        timestamp = str(int(time.time() * 1000))
        token_part = self.access_token or ""
        sign_source = f"{self.access_id}{token_part}{timestamp}{string_to_sign}"
        signature = (
            hmac.new(self.access_secret, sign_source.encode(), hashlib.sha256).hexdigest().upper()
        )
        headers = {
            "client_id": self.access_id,
            "sign": signature,
            "t": timestamp,
            "sign_method": "HMAC-SHA256",
            "Content-Type": "application/json",
        }
        if self.access_token:
            headers["access_token"] = self.access_token
        return headers

    def request(
        self,
        method: str,
        path: str,
        params: dict[str, Any] | None = None,
        body: Any | None = None,
        raise_on_error: bool = True,
    ) -> Any:
        self._config.debug(f"cloud {method} {path}")
        response = requests.request(
            method.upper(),
            f"{self.endpoint}{path}",
            params=params,
            data=self._body_text(body) if body is not None else None,
            headers=self._sign(method, path, params, body),
            timeout=30,
        )
        try:
            payload = response.json()
        except ValueError:
            response.raise_for_status()
            return response.text
        if raise_on_error and (not response.ok or payload.get("success") is False):
            raise RuntimeError(
                f"Tuya API {method} {path} failed: HTTP {response.status_code}: {payload}"
            )
        return payload

    def connect(self) -> None:
        """Obtain an access token from the Tuya token endpoint."""
        self._config.debug("cloud: obtaining access token")
        payload = self.request("GET", "/v1.0/token", {"grant_type": 1})
        token = payload.get("result", {}).get("access_token")
        if not token:
            raise RuntimeError(f"Tuya token response did not contain an access_token: {payload}")
        self.access_token = token

    def status(self) -> dict[str, Any]:
        """Fetch current device status DPs via the Cloud API."""
        self.connect()
        payload = self.request(
            "GET",
            f"/v1.0/iot-03/devices/{self.device_id}/status",
            raise_on_error=False,
        )
        if payload.get("success") is not True:
            payload = self.request("GET", f"/v1.0/devices/{self.device_id}/status")
        rows = payload.get("result") or []
        return {row.get("code"): row.get("value") for row in rows if isinstance(row, dict)}

    def properties(self) -> dict[str, dict[str, Any]]:
        """Fetch shadow properties (includes memory_height and preset data)."""
        self.connect()
        payload = self.request("GET", f"/v2.0/cloud/thing/{self.device_id}/shadow/properties")
        rows = payload.get("result", {}).get("properties", [])
        return {row.get("code"): row for row in rows if isinstance(row, dict) and row.get("code")}

    def send(self, commands: list[dict[str, Any]]) -> Any:
        """Send commands to the device, falling back to the legacy endpoint."""
        self.connect()
        body = {"commands": commands}
        payload = self.request(
            "POST",
            f"/v1.0/iot-03/devices/{self.device_id}/commands",
            body=body,
            raise_on_error=False,
        )
        if payload.get("success") is True:
            return payload
        legacy = self.request(
            "POST",
            f"/v1.0/devices/{self.device_id}/commands",
            body=body,
            raise_on_error=False,
        )
        if legacy.get("success") is not True:
            raise RuntimeError(f"Tuya command failed: iot-03={payload}; legacy={legacy}")
        return legacy

File: internal/tuya/cloud.go
package tuya

import (
	"bytes"
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/yashiels/orro-cli/internal/config"
)

// Command is a single DP code + value pair used by both LAN and Cloud.
type Command struct {
	Code  string `json:"code"`
	Value any    `json:"value"`
}

// Cloud is a Tuya Cloud API client with HMAC-SHA256 request signing.
type Cloud struct {
	cfg         *config.Config
	endpoint    string
	clientID    string
	secret      []byte
	deviceID    string
	accessToken string
	http        *http.Client
}

// NewCloud creates a Cloud client. Connect() must be called before any device operations.
func NewCloud(cfg *config.Config) *Cloud {
	return &Cloud{
		cfg:      cfg,
		endpoint: strings.TrimRight(cfg.Endpoint, "/"),
		clientID: cfg.AccessID,
		secret:   []byte(cfg.AccessSecret),
		deviceID: cfg.DeviceID,
		http:     &http.Client{Timeout: 30 * time.Second},
	}
}

// Connect obtains an access token from the Tuya token endpoint.
func (c *Cloud) Connect() error {
	c.cfg.Debug("cloud: obtaining access token")
	payload, err := c.request("GET", "/v1.0/token", url.Values{"grant_type": {"1"}}, nil, false)
	if err != nil {
		return fmt.Errorf("cloud connect: %w", err)
	}
	result, ok := payload["result"].(map[string]any)
	if !ok {
		return fmt.Errorf("cloud connect: unexpected response: %v", payload)
	}
	token, _ := result["access_token"].(string)
	if token == "" {
		return fmt.Errorf("cloud connect: no access_token in response: %v", payload)
	}
	c.accessToken = token
	c.cfg.Debug("cloud: token obtained")
	return nil
}

// Status fetches current device DP values.
func (c *Cloud) Status() (map[string]any, error) {
	if err := c.Connect(); err != nil {
		return nil, err
	}

	// Try iot-03 endpoint first, fall back to legacy.
	payload, err := c.request("GET",
		fmt.Sprintf("/v1.0/iot-03/devices/%s/status", c.deviceID),
		nil, nil, false)
	if err != nil || !truthy(payload["success"]) {
		payload, err = c.request("GET",
			fmt.Sprintf("/v1.0/devices/%s/status", c.deviceID),
			nil, nil, true)
		if err != nil {
			return nil, fmt.Errorf("cloud status: %w", err)
		}
	}

	rows, _ := payload["result"].([]any)
	out := make(map[string]any, len(rows))
	for _, row := range rows {
		if m, ok := row.(map[string]any); ok {
			code, _ := m["code"].(string)
			if code != "" {
				out[code] = m["value"]
			}
		}
	}
	return out, nil
}

// Properties fetches shadow properties (includes memory_height preset data).
func (c *Cloud) Properties() (map[string]map[string]any, error) {
	if err := c.Connect(); err != nil {
		return nil, err
	}

	payload, err := c.request("GET",
		fmt.Sprintf("/v2.0/cloud/thing/%s/shadow/properties", c.deviceID),
		nil, nil, true)
	if err != nil {
		return nil, fmt.Errorf("cloud properties: %w", err)
	}

	result, _ := payload["result"].(map[string]any)
	rows, _ := result["properties"].([]any)
	out := make(map[string]map[string]any, len(rows))
	for _, row := range rows {
		if m, ok := row.(map[string]any); ok {
			code, _ := m["code"].(string)
			if code != "" {
				out[code] = m
			}
		}
	}
	return out, nil
}

// Send sends device commands via the Cloud API.
func (c *Cloud) Send(commands []Command) (map[string]any, error) {
	if err := c.Connect(); err != nil {
		return nil, err
	}

	body := map[string]any{"commands": commands}

	// Try iot-03 first.
	payload, err := c.request("POST",
		fmt.Sprintf("/v1.0/iot-03/devices/%s/commands", c.deviceID),
		nil, body, false)
	if err == nil && truthy(payload["success"]) {
		return payload, nil
	}

	// Fall back to legacy endpoint.
	legacy, err2 := c.request("POST",
		fmt.Sprintf("/v1.0/devices/%s/commands", c.deviceID),
		nil, body, false)
	if err2 != nil {
		return nil, fmt.Errorf("cloud send failed: iot-03=%v; legacy=%v", err, err2)
	}
	if !truthy(legacy["success"]) {
		return nil, fmt.Errorf("cloud send: iot-03=%v; legacy=%v", payload, legacy)
	}
	return legacy, nil
}

// ──────────────────────────────────────────────
// HTTP + signing
// ──────────────────────────────────────────────

func (c *Cloud) request(
	method, path string,
	params url.Values,
	body any,
	failOnError bool,
) (map[string]any, error) {
	c.cfg.Debug(fmt.Sprintf("cloud: %s %s", method, path))

	var bodyBytes []byte
	if body != nil {
		var err error
		bodyBytes, err = json.Marshal(body)
		if err != nil {
			return nil, err
		}
	}

	// Build query string.
	query := canonicalQuery(params)
	fullPath := path
	if query != "" {
		fullPath = path + "?" + query
	}

	headers := c.sign(method, path, query, bodyBytes)

	reqURL := c.endpoint + fullPath
	var reqBody io.Reader
	if len(bodyBytes) > 0 {
		reqBody = bytes.NewReader(bodyBytes)
	}

	req, err := http.NewRequestWithContext(context.Background(), method, reqURL, reqBody)
	if err != nil {
		return nil, err
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := c.http.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result map[string]any
	if err := json.Unmarshal(respBody, &result); err != nil {
		if failOnError {
			return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, respBody)
		}
		return nil, nil
	}

	if failOnError && (resp.StatusCode < 200 || resp.StatusCode >= 300 || !truthy(result["success"])) {
		return nil, fmt.Errorf("API %s %s failed: HTTP %d: %v", method, path, resp.StatusCode, result)
	}

	return result, nil
}

func (c *Cloud) sign(method, path, query string, bodyBytes []byte) map[string]string {
	bodyHash := sha256Hex(bodyBytes)
	urlPath := path
	if query != "" {
		urlPath = path + "?" + query
	}

	stringToSign := strings.Join([]string{
		strings.ToUpper(method),
		bodyHash,
		"",
		urlPath,
	}, "\n")

	ts := strconv.FormatInt(time.Now().UnixMilli(), 10)
	tokenPart := c.accessToken

	signSource := c.clientID + tokenPart + ts + stringToSign

	h := hmac.New(sha256.New, c.secret)
	h.Write([]byte(signSource))
	signature := strings.ToUpper(hex.EncodeToString(h.Sum(nil)))

	headers := map[string]string{
		"client_id":    c.clientID,
		"sign":         signature,
		"t":            ts,
		"sign_method":  "HMAC-SHA256",
		"Content-Type": "application/json",
	}
	if c.accessToken != "" {
		headers["access_token"] = c.accessToken
	}
	return headers
}

// canonicalQuery sorts and URL-encodes query parameters for Tuya signing.
func canonicalQuery(params url.Values) string {
	if len(params) == 0 {
		return ""
	}
	keys := make([]string, 0, len(params))
	for k := range params {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	var pairs []string
	for _, k := range keys {
		for _, v := range params[k] {
			pairs = append(pairs, url.QueryEscape(k)+"="+url.QueryEscape(v))
		}
	}
	return strings.Join(pairs, "&")
}

func sha256Hex(data []byte) string {
	h := sha256.Sum256(data)
	return hex.EncodeToString(h[:])
}

func truthy(v any) bool {
	if v == nil {
		return false
	}
	if b, ok := v.(bool); ok {
		return b
	}
	return false
}
